### ISSUE TITLE
PDF can not open in Adobe InDesign

### ISSUE THREAD

**Issue by Kishanpatel96** [pymupdf/PyMuPDF#958]
_**Please provide all mandatory information!**_
I am using PyMuPDF for writing a text on PDF. I am successfully able to write text on PDF with this library. If I open PDF normally or Other Adobe software then It is working fine.
## Describe the bug (mandatory)
If I open an output PDF file in Adobe InDesign then it is giving an error that PDF cannot Open.
If I am not writing text and Simply save PDF file then also it is open in Adobe InDesign 
## To Reproduce (mandatory)
Explain the steps to reproduce the behavior, For example, include a minimal code snippet, example files, etc.
I have tried multiple methods to write text on PDF.
**1. With insertTextbox method.**
First way or the regular way I am trying to write text on PDF using this code.

```
pdf.insertTextbox(position,
                word,
                align=alignment,
                fontname=fontname,
                fontfile=slot_data['fontFile'],
                fontsize=slot_data['fontSize'],
                rotate=0,
                color=RGBColor,
                )

pdf_document.save(output_file,garbage=3, deflate=True, pretty=True)
```
## Expected behavior (optional)
Describe what you expected to happen (if not obvious).
I want to open a PDF file the normal way in Adobe InDesign. 
Can you tell me it is possible or not. If possible then can you guide me on how I can do it?
## Your configuration (mandatory)
 - Operating system, potentially version and bitness
 - Python version, bitness
 - PyMuPDF version, installation method (**wheel** or **generated** from source).

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

## Additional context (optional)
Thank you,
Kishan Patel

**Comment by JorjMcKie** [comment]
Problem not reproducible like that.
Please provide the full script or confirm that your `insert_textbox()` indeed uses the correct call pattern:
position = a rectangle?
word = a string or list of strings
etc.
Which fonts are you using?
Also an example PDF.

**Comment by JorjMcKie** [comment]
Ah and of course the rest of mandatory information: your environment and PyMuPDF version.

**Comment by Kishanpatel96** [comment]
PyMuPDF version is 1.18.6.
Python version is 3.6 
Flask version is 1.0.2


```
position = fitz.Rect(50,200,150,300)
        print_siegel_service.siegel_file.insert_textbox(position,
        'Hello',
        align=0,
        fontname="F!",
        fontfile="font/Nobel-Regular.otf",
        fontsize=15,
        rotate=0,
        color=(0,0,0),
        )
```

Yes as you told I have tried with insert_textbox() still PDF is not opening in Adobe InDesign.

**Comment by JorjMcKie** [comment]
operating system?

**Comment by Kishanpatel96** [comment]
I am using macOS Mojave 10.14.6 Operating System.
My Project is running on Docker and the version of Docker is 2.3

I hope this is enough information.

**Comment by JorjMcKie** [comment]
How do you save the modified PDF: garbage options?, deflate=True?, `page.clean_contents()` after text insertion ...

Adobe InDesign is not a free product - it costs money - and I am not using it, nor will I buy it.
There is a grace period of 7 days after downloading a test version. So the issue must be resolved within a week ...

So before using up this precious timeframe:
Can you confirm, that a PDF with your inserted text causes no problems when opening it in Adobe Acrobat and all other PDF viewers you may be using? Outside docker, on your desktop?

**Comment by Kishanpatel96** [comment]
```
file_path = "filename.pdf"
pdf_document = fitz.open(file_path)
page = pdf_document[0]

position = fitz.Rect(50,200,150,300)
page.insert_textbox(position,
'Hello',
align=0,
fontname="F1",
fontfile="font/Nobel-Regular.otf",
fontsize=15,
rotate=0,
color=(0,0,0),
)
page.clean_contents()

pdf_document.save('outputfile.pdf', garbage=2, deflate=True, pretty=True)
```

I am using this code to write text on PDF. I have tried garbage values from 1 to 4 but the output is the same (Cannot open PDF in Adobe InDesign).

Yes, I am able to open a PDF file in Adobe Acrobat and Adobe illustrate also in google Crome and normal PDF viewers.
But the problem is coming in Adobe InDesign also.

If you have other code suggestions then also you can tell me I will try here and check also.

Yes I am also want to solve this issue soon.

**Comment by JorjMcKie** [comment]
Your code looks good! And thanks for the accompanying info.
Ok, I will now go ahead and check with InDesign.

**Comment by Kishanpatel96** [comment]
Okay, I am really waiting for your answer.
Please make sure you have an open PDF (Not Place) in Adobe InDesign.

I hope we can solve this issue soon.

**Comment by JorjMcKie** [comment]
Installed Adobe InDesign.
It does not open **_any_** PDF! What is wrong?

**Comment by Kishanpatel96** [comment]
Do you mean printed PDF from PyMuPDF or normal PDF?

And one more time.
You don't have to open PDF in Adobe InDesign.

You have to delete Place from the menu.
For Reference, you can use [this video](https://www.youtube.com/watch?v=qXFTqEZm-x0&feature=emb_rel_pause).

**Comment by JorjMcKie** [comment]
It works:
Took an existing PDF and did this:
```python
>>> doc=fitz.open("v110-changes.pdf")
>>> page=doc[0]
>>> text = "** test ** test ** test ** test ** test ** test **"
>>> page.insert_textbox(page.rect, text)
830.0949823856354
>>> doc.save("x.pdf")
>>> 
```
Then placed `x.pdf` in an indesign document:
![grafik](https://user-images.githubusercontent.com/8290722/111665162-e0e1f800-87e8-11eb-99a1-afed86f03eb3.png)

Same thing with a PDF create new in PyMuPDF - no issues.

**Comment by Kishanpatel96** [comment]
Can you tell me your full configuration? What version you are using for all software including InDesign?

**Comment by JorjMcKie** [comment]
Windows 10 64bit, Python 3.9 64bit, InDesign 16.1, PyMuPDF v1.18.8

By the way - as soon as I use a **_fontfile-based_** font, InDesign refuses to open the PDF.
Choosing one of the Base-14 font (Helvetica, Times-Roman or Courier and their respective font weights) all works fine.

**Comment by Kishanpatel96** [comment]
Yes, you are right I am also able to open PDF in InDesign if I am using a font from Base-14 fonts.

Can you give me any idea why it is not working with **fontfile-based** font?
And it could be possible to solve with fontfile or something?

Because I want to use fontfile.

**Comment by JorjMcKie** [comment]
The problem is larger than around fontfile. And it is **_InDesign-specific_**.
Even if a convert a previously accepted PDF to a new PDF using `mutool draw -o test.pdf v110-changes.pdf` then the new file `test.pdf` will again not be accepted.

So in summary, InDesign only accepts some subset of otherwise perfectly valid PDFs -- PDFs that can be opened in Adobe Acrobat and all other PDF viewers without restrictions

I have no idea how this comes about and probably must be addressed / interrogated by InDesign. For you immediate needs you may want to convert your modified PDF to some image file -- page by page -- and them import these.
I see no way to help. 
And it definitely does not look like a PyMuPDF or MuPDF error.

**Comment by Kishanpatel96** [comment]
Yes, I got your point but Can you give me a way to make my font vectorized while writing it on PDF or on Saving time?

**Comment by JorjMcKie** [comment]
> font vectorized

Don't understand that.

You could convert a PDF page to an SVG image and then import this one. This is a vector format and by defaut font-independent.

**Comment by Kishanpatel96** [comment]
So for your clear understanding.
I want to write text on PDF. And If I write text as we are doing currently then it can change with Software like Adobe or and others.

So I want to make the text which will be written by PyMuPDF will be not changed by any software.
So with making the font vectorized it become Shape and it will be not changeable.

So is there any way to make text not editable by other software?

**Comment by JorjMcKie** [comment]
> So is there any way to make text not editable by other software?

Well, you can always password-protect a PDF and encrypt it such, that it can be read, but not changed by anyone without the password.

What you meant amounts to **_not_** writing the text by using a font. Instead paint each character by elementary PDF drawing commands like a capital letter "D" is painted as a closed semi-circle.
This is not (directly) supported by MuPDF. But the default SVG output (`text_as_path=True`) does exactly that as explained above.

**Comment by JorjMcKie** [comment]
BTW: the correct term is "text as path". Text written using a font, by definition is vectorized.

**Comment by JorjMcKie** [comment]
> I want to write text on PDF. And If I write text as we are doing currently then it can change or edit with Software like Adobe or and others.

Changing text no matter what you do is always possible. Even when written as paths.
Use an OCR software and re-insert the interpreted, but modified text.

**Comment by Kishanpatel96** [comment]
I feel **Well, you can always password-protect a PDF and encrypt it such, that it can be read, but not changed by anyone without the password.** this is a good way to go and I want to try it.

Can you give me an example code for this feature? Because we have permission, user_pw, and owner_pw
So I will example code will be good.

**Comment by JorjMcKie** [comment]
Use an owner password and encrypt the file with a method of your liking.
If you **_omit_** the user password, anyone can open the document without entering a password and will automatically be identified as a user - otherwise opening requires a password.
After opening, the actions allowed are controlled by the `permissions` integer - which is a bit-field actually.
To configure the permissions do this
```python
perm = fitz.PDF_PERM_ACCESSIBILITY  # always start with this
perm |= fitz.PDF_PERM_PRINT  # then OR additional allowed actions into it
perm |= ...
doc.save("encrypted.pdf", owner_pw="owner", user_pw="user", permissions=perm, encryption=fitz.PDF_ENCRYPT_AES_256)
```
The maximum password length is 50 bytes.
When opening requires a password, the entered password value will automatically identify you as owner or user (or even both if the passwords are equal).
